doc/smart-switch/high-availability/smart-switch-ha-hld.md

```diff
@@ -0,0 +1,2208 @@
+3. The outer packet ***MUST*** have the same DSCP fields as the inner packet.
+
+The detailed data path is shown below.
+
+##### 4.2.2.1. Tunneling packet to local DPU
+
+<p align="center"><img alt="Tunneling packet to local DPU" src="./images/npu-to-dpu-tunnel-local.svg"></p>
+
+##### 4.2.2.2. Tunneling packet to remote DPU
+
+<p align="center"><img alt="Tunneling packet to remote DPU" src="./images/npu-to-dpu-tunnel-remote.svg"></p>
+ 
+##### 4.2.2.3. Tunneling packet between DPUs
+
+This tunnel needs to be also implemented by DPU, because in planned or unplanned events, traffic might land on the in non-active side DPU and needs to be tunneled. This is using the same NPU-to-DPU traffic tunnel.
+ 
+<p align="center"><img alt="Tunneling packet between DPUs" src="./images/npu-to-dpu-tunnel-from-dpu.svg"></p>
+
+##### 4.2.2.4. Put all tunnels together
+
+Now let’s put all tunnels together. This can happen when the packets land on other switches that help traffic forwarding in the middle of HA state transition.
+ 
+<p align="center"><img alt="Tunneling packet between DPUs" src="./images/npu-to-dpu-tunnel-2-stages.svg"></p>
+
+#### 4.2.3. DPU-to-DPU data plane channel
+
+This channel is implemented below SAI, used by inline flow replication or any other network-triggered inline data sync. Since traffic buffer on switch / DPU is limited, we cannot hold the packet when doing the flow replication, hence data sync between DPUs such as flow replication needs to be done inline.
+
+###### 4.2.3.0.1. Data plane channel data path
+
+Data plane channel is essentially a UDP-based tunnel between DPUs, used in inline flow replications. Conceptually, it works as below in high-level:
+
+1. The outer packet ***MUST*** use the “DP channel port” specified via SONiC.
+2. The outer packet ***MUST*** calculate the source port from the 5 tuples of the inner packet and fold it into a port range specified by SONiC.
+3. The outer packet ***MUST*** have the same DSCP fields as the inner packet.
+4. Different vendors ***CAN*** have their own definition of metadata format as payload.
+
+Here is an example from "[Flow replication data path overview – Data plane sync data path](#1221-data-plane-sync-channel-data-path)":
+
+<p align="center"><img alt="Flow replication data plane sync data path" src="./images/flow-replication-data-path.svg"></p>
+
+###### 4.2.3.0.2. Multi-path data plane channel
+
+Since data plane channel is inline, it is very important to make sure whenever gray network failure happens (traffic being blackholed partially), we should avoid causing the traffic for the entire ENI to be dropped.
+
+To provide HA for DPU-to-DPU communication, we can also leverage our ECMP network setup. When sending the tunneled packet, we must calculate the source port from the 5 tuples of the inner packet and fold it into a port range specified by SONiC. This helps us to spread the packets across all possible paths and avoid the blackhole problem.
```

Comment:
>we must calculate the source port from the 5 tuples of the inner packet and fold it into a port range specified by SONiC

Do you mean "The source port is one of the 5 members of the inner packet's 5-tuple and as such it can be chosen to ensure the egress ports are evenly used?" I'm also considering the next section's language:
>The algorithm for port selection doesn’t matter. As long as we can spread the packet across all ports, we should be good.

My concern is, if typical ECMP hash-based techniques are used (hash on a 5-tuple, modulo a list of egress ports), it's not straightforward to predict the resultant egress port for a given 5-tuple (it can vary across different ASIC implementations), and it is even less straightforward to ensure all ports are evenly chosen for a set of flows. (ECMP Polarization is one malignant side-effect of this.) So I'm not sure how one would "choose an algorithm" and derive a list of source ports to plug into these packets to effect balanced use of egress ports.

Reply:
oops, the second one is missed when updating the requirements. will get it updated too.

Reply:
for the ECMP polarization issue, you are right. Hash-based ECMP will never be a perfect or even great solution for elephant flows or determine the data path of each flow, but it should be a good starting point for us. This is because the number of flows we are handling is large enough and we have enough number of links between each T1, so as long as they can be spread, some skews will be acceptable. And it is hard to determine the data path of each flows anyway (vendor implementation varies and link states changes overtime as well), so we rarely use the 5-tuple to query the next hop. :D

Reply:
The doc is updated now.

Reply:
please let me know, if still have any concerns. :D

Reply:
>This is because the number of flows we are handling is large enough and we have enough number of links between each T1, so as long as they can be spread, some skews will be acceptable.

I'm still unclear how you'll achieve effective egress port spreading, with any confidence, based on some list of source ports. It sounds like "spray and pray" (hah hah I just thought of that).

Reply:
XD, kinda yes and no at the same time. From DPU perspective, it is indeed "spray and pray". However:

- For network failures, the network itself is heavily monitored. So, when any issue happens, it can isolate itself and shift the traffic.
- For load spreading, the number links between T1s are much higher than the number of DPUs in the switch, so although plain hash-based ECMP might not sound beautiful, but the amount of skews are acceptable.

Reply:
OK, thanks for explaining!

Reply:
thank-you for pointing it out!